from snakemake.utils import min_version

min_version("7.3.6")



### workflow.basedir for the location of the Snakefile
configfile: os.path.join(workflow.basedir, "..", "config/config.yaml")


report: "report/workflow.rst"


print(os.getcwd())


# Using singularity for conda/mamba support
# container: "docker://condaforge/mambaforge"

include: "rules/common.smk"
include: "rules/pedigree.smk"
include: "rules/ref.smk"

if config["fastq_input"]["enable"]:
    include: "rules/premap.smk"
    include: "rules/mapping.smk"

include: "rules/bam_qc.smk"
include: "rules/gatk_hc.smk"
include: "rules/deepvariant.smk"
include: "rules/call_dnm.smk"
include: "rules/jigv.smk"

rule all: 
    input: 
        rules.bwa_index.output,
        genome_dict,
        expand(output_dir + "/fq2bam/{SAMPLE_ID}.bam", SAMPLE_ID=subjs ),
        expand(output_dir + "/gatk_cgp/{fam}.cgp_norm.vcf.gz", fam=fam_ids),
        expand(output_dir + "/dnm_vcf/{fam}.dnm.vcf.gz", fam=fam_ids),
        expand(output_dir + "/call_JIGV/{fam}.JIGV.html", fam=fam_ids),
        optional_output=optional_output,
        


